db/background_mgr_attachment_migration.go

```diff
@@ -0,0 +1,288 @@
+
+		// Don't want to process raw binary docs
+		// The binary check should suffice but for additional safety also check for empty bodies
+		if event.DataType == base.MemcachedDataTypeRaw || len(event.Value) == 0 {
+			return true
+		}
+
+		// We only want to process full docs. Not any sync docs.
+		if strings.HasPrefix(docID, base.SyncDocPrefix) {
+			return true
+		}
+
+		a.DocsProcessed.Add(1)
+		syncData, rawBody, rawXattrs, err := UnmarshalDocumentSyncDataFromFeed(event.Value, event.DataType, collection.userXattrKey(), false)
+		if err != nil {
+			failProcess(err, "error unmarshaling document %s: %v", base.UD(docID), err)
+		}
+
+		if rawXattrs[base.GlobalXattrName] != nil {
+			// global xattr already defined for this document, don't process
+			return true
+		}
+
+		if syncData == nil || syncData.Attachments == nil {
+			// no attachments to migrate
+			return true
+		}
+
+		collCtx := collection.AddCollectionContext(ctx)
+		collWithUser := &DatabaseCollectionWithUser{
+			DatabaseCollection: collection,
+		}
+		if event.DataType&base.MemcachedDataTypeXattr == 0 {
+			// no xattrs on dcp event we have inline sync xattr to migrate
+			opts := DefaultMutateInOpts()
+			existingDoc := &sgbucket.BucketDocument{
+				Cas:    event.Cas,
+				Body:   rawBody,
+				Expiry: event.Expiry,
+			}
+			_, _, err = collWithUser.migrateMetadata(ctx, docID, existingDoc, opts)
+			if err != nil {
+				failProcess(err, "error migrating document attachment metadata for doc: %s: %v", base.UD(docID), err)
+			} else {
+				a.DocsChanged.Add(1)
+			}
+		} else {
+			// normal xattr migration to take place
+			err = collWithUser.MigrateAttachmentMetadata(collCtx, docID, event.Cas, syncData)
+			if err != nil {
+				failProcess(err, "error migrating document attachment metadata for doc: %s: %v", base.UD(docID), err)
+			} else {
+				a.DocsChanged.Add(1)
+			}
+		}
+		return true
+	}
+
+	bucket, err := base.AsGocbV2Bucket(db.Bucket)
+	if err != nil {
+		return err
+	}
+
+	collectionIDs := getCollectionIDs(db)
+	dcpOptions := getMigrationDCPClientOptions(collectionIDs, db.Options.GroupID, db.MetadataKeys.DCPCheckpointPrefix(db.Options.GroupID))
+	dcpFeedKey := GenerateAttachmentMigrationDCPStreamName(a.MigrationID)
+	dcpClient, err := base.NewDCPClient(ctx, dcpFeedKey, callback, *dcpOptions, bucket)
+	if err != nil {
+		base.WarnfCtx(ctx, "[%s] Failed to create attachment migration DCP client: %v", migrationLoggingID, err)
+		return err
+	}
+	base.InfofCtx(ctx, base.KeyAll, "[%s] Starting DCP feed %q for attachment migration", migrationLoggingID, dcpFeedKey)
+
+	doneChan, err := dcpClient.Start()
+	if err != nil {
+		base.WarnfCtx(ctx, "[%s] Failed to start attachment migration DCP feed: %v", migrationLoggingID, err)
+		_ = dcpClient.Close()
+		return err
+	}
+	base.DebugfCtx(ctx, base.KeyAll, "[%s] DCP client started.", migrationLoggingID)
```

Comment:
Including something about attachment migration in this log message will avoid future confusion.